de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerToolBarController.java

```diff
@@ -0,0 +1,194 @@
+package de.tu_bs.cs.isf.e4cf.parts.project_explorer.controller;
+
+
+import org.eclipse.swt.widgets.Shell;
+
+import de.tu_bs.cs.isf.e4cf.core.stringtable.E4CEventTable;
+import de.tu_bs.cs.isf.e4cf.core.util.ServiceContainer;
+import de.tu_bs.cs.isf.e4cf.parts.project_explorer.handlers.FileImportHandler;
+import de.tu_bs.cs.isf.e4cf.parts.project_explorer.handlers.NewFolderHandler;
+import de.tu_bs.cs.isf.e4cf.parts.project_explorer.handlers.RemoveFileCommand;
+import de.tu_bs.cs.isf.e4cf.parts.project_explorer.handlers.ShowInExplorerHandler;
+import de.tu_bs.cs.isf.e4cf.parts.project_explorer.stringtable.FileTable;
+import javafx.event.EventHandler;
+import javafx.event.ActionEvent;
+import javafx.scene.control.Button;
+import javafx.scene.control.Separator;
+import javafx.scene.control.TextField;
+import javafx.scene.control.ToolBar;
+import javafx.scene.control.Tooltip;
+import javafx.scene.image.ImageView;
+
+/**
+ * Fills and handles the ProjectExplorer Toolbar.
+ * Please note that the use of FX Accelerators as shortcuts is discouraged as FX <13 consumes their KeyCombos application wide
+ * and this is not desired behavior.
+ */
+public class ProjectExplorerToolBarController {
+
+	private ServiceContainer services;
+
+	private ToolBar bar;
+
+	private Button btnUndo;
+	private Button btnRedo;
+	private Button btnNewFolder;
+	private Button btnImportFiles;
+	private Button btnDelete;
+	private Button btnShowExplorer;
+	boolean isFlatView = false;
+
+	/**
+	 * Fills the toolbar with Buttons performing project explorer related actions.
+	 * @param toolbar The FXML Toolbar Object defined in UI
+	 * @param services The ServiceContainer that holds all e4c Services, required for some handlers
+	 * @param shell The current UI Shell passed on from the ViewController's Parent, required for some handlers
+	 */
+	public ProjectExplorerToolBarController(ToolBar toolbar, ServiceContainer services, Shell shell) {
+		bar = toolbar;
+		this.services = services;
+
+		// Search Section
+		TextField search = new TextField();
+		search.setVisible(false);
+		search.setOnAction(actionEvent -> {
+			services.eventBroker.send(E4CEventTable.EVENT_FILTER_CHANGED, search.getText());
+		});
+
+		ImageView searchImage = services.imageService.getFXImage(null, FileTable.SEARCH_PNG);
+		Tooltip searchTooltip = new Tooltip("Search Files");
+		ImageView xImage = services.imageService.getFXImage(null, FileTable.X_PNG);
+		Tooltip closeSearchTooltip = new Tooltip("Exit Search");
+		Button searchButton = new Button("", searchImage);
+		searchButton.setTooltip(searchTooltip);
+		searchButton.setOnAction(actionEvent -> {
+			search.setVisible(!search.isVisible());
+			// Removing and readding the search bar allows automatic resizing
+			if (search.isVisible()) {
+				searchButton.setGraphic(xImage);
+				searchButton.setTooltip(closeSearchTooltip);
+				bar.getItems().add(1, search);
```

Comment:
We could add that as soon as the searchbar(textfield) is visible it also gets the focus so the user can directly start typing.